### PR TITLE
[feature/bots] Add CI failure bot for automated responses #524

### PR DESCRIPTION
**Pull request by Srinath0916**
Implements automated bot responses for common CI failures:

Analyzes QA check failures and provides OpenWISP-aligned guidance
Detects test failures and setup issues
Provides specific fix suggestions with links
Posts helpful guidance comments automatically

## Checklist
- [x] I have read the OpenWISP Contributing Guidelines
- [x] I have manually tested the changes proposed in this pull request
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code
- [x] I have updated the documentation

## Reference to Existing Issue
Addresses #524

## Description of Changes
This PR introduces a CI Failure Bot that automatically analyzes failed CI builds and posts helpful guidance comments on pull requests.

What the bot does:

Analyzes CI failure logs and provides factual diagnosis
Posts a single deduplicated comment with clear next steps
Always recommends the official OpenWISP QA workflow commands
Never suggests raw linters (black, flake8, isort)
Gracefully handles edge cases (missing API keys, fork PRs, etc.)
Skips dependabot PRs automatically
Recommended commands (always consistent):

pip install -e .[qa]
openwisp-qa-format
./run-qa-checks

## Demo repository & real-world testing

The CI failure bot has been tested using a dedicated demo repository:

https://github.com/srinath-testing/gemini-ci-bot-demo

The following real-world failure scenarios are covered using actual pull requests:

Demo PR  – QA / formatting failures → https://github.com/srinath-testing/gemini-ci-bot-demo/pull/14
Demo PR  – Unit test failures → https://github.com/srinath-testing/gemini-ci-bot-demo/pull/15
Demo PR  – Mixed issues (imports + formatting) → https://github.com/srinath-testing/gemini-ci-bot-demo/pull/16

Each example demonstrates:

CI failure occurs
Bot posts automatic comment
Correct OpenWISP-aligned guidance provided
No duplicate or misleading output

Notes
The bot is intentionally designed as a conservative first-line helper and does not attempt perfect root-cause classification in all complex scenarios. It focuses on reliability and consistency with OpenWISP standards.

AI integration (Gemini) is optional and never blocks posting feedback to contributors.

### CODE REVIEW

File: .github/scripts/ci_failure_bot.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+## Setup Failed
+
+There was an issue with dependencies or environment setup.
+
+Check if you added new dependencies to requirements-test.txt.
+Verify Python/Django version compatibility:
+- Python: 3.10, 3.11, 3.12, 3.13
+- Django: 4.2, 5.0, 5.1, 5.2
+
+The CI will retry automatically, or push a small change to trigger rebuild.
```
Comment (nemesifier) [inline .github/scripts/ci_failure_bot.py]: I don't understand the point of this. Given my previous comment, do you think this is still needed?

**Review by nemesifier — CHANGES_REQUESTED** [review]
I am not seeing any response in https://github.com/Srinath0916/ci-bot-test/pull/1. Is it really working?

Would this work as a reusable github actions?

Docs and tests are completely missing but I guess we can spend time on that once we are sure this is going in the right direction.

PS: please let's exclude dependabot PRs.

File: .github/workflows/ci-failure-bot.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        run: |
+          pip install requests PyGithub
+
+      - name: Run CI Failure Bot
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          WORKFLOW_RUN_ID: ${{ github.event.workflow_run.id }}
+          REPOSITORY: ${{ github.repository }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number || '' }}
+        run: python .github/scripts/ci_failure_bot.py
```
[the remark this answers is not in the document]
Comment (Srinath0916) [reply .github/workflows/ci-failure-bot.yml]: Thanks @nemesifier , for pointing this out,  I should have clarified this earlier.

I double-checked the workflow logs: the bot is triggering correctly, detecting the failure type, and attempting to post a comment. In my personal test repository, the comment fails with a 403 Resource not accessible by integration, which is a known limitation when using workflow_run + GITHUB_TOKEN on personal repos. The workflow log reflects this.

In OpenWISP organization repositories, where the token has issue/PR write permissions, this limitation should not apply.
 
Regarding reuse: Yes, the logic can be made reusable across OpenWISP repositories by parameterizing repository-specific context (repo name, URLs, workflow name). Other repositories would only need to add the workflow file and rely on org-level secrets.
I’ll also exclude Dependabot PRs at both the workflow and script level.

File: .github/scripts/ci_failure_bot.py (hunk 1 above)
Comment (Srinath0916) [reply .github/scripts/ci_failure_bot.py]: This static message approach isn't needed given the Gemini integration direction.

The static responses were just to validate the workflow mechanics, but the AI analysis will provide much better, context-specific guidance.

I'll focus on the Gemini approach instead.

**Review by nemesifier — COMMENTED** [review]
@Srinath0916 before merging it here we need to test this somewhere else.

I do not see any output in https://github.com/Srinath0916/ci-bot-test/pull/1, how are you testing this?

**Comment by Srinath0916** [comment]
Hi @nemesifier ,

I can show you the working bot demo here: https://github.com/srinath-testing/gemini-ci-bot-demo/pull/1

You can see the bot posted: "🤖 Demo CI Failure Bot (AI-powered)" with structured analysis.

The reason you don't see output in Srinath0916/ci-bot-test#1 is that personal repos have token permission limitations with workflow_run triggers. The demo repo I created specifically for this has the right setup.

The Gemini integration is ready locally - I can push it once you confirm the demo proves the concept works. Would you like me to proceed with the full AI implementation?

**Comment by Srinath0916** [comment]
@nemesifier Gemini implementation is ready.  Let me know if you'd like any adjustments before I proceed working on tests and docs next

File: .github/scripts/ci_failure_bot.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env python3
+"""
+CI Failure Bot - AI-powered analysis of build failures using Gemini
+"""
+
+import json
```
Comment (nemesifier) [inline .github/scripts/ci_failure_bot.py]: What happens if a contributor breaks this logic? If we don't add any automated test at all for this, it could stop working any time and we wouldn't even know until the change is merged to master and new PRs will be opened. We need some tests for this that run only in openwisp-utils.

Comment (Srinath0916) [reply .github/scripts/ci_failure_bot.py]: @nemesifier , Added comprehensive documentation following the structure you linked and a full test suite that covers all the bot functions. Now if someone breaks the logic, the tests will catch it before it reaches master.

File: .github/workflows/ci-failure-bot.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+          pip install requests>=2.31.0 PyGithub>=2.0.0 google-generativeai>=0.3.0
+
+      - name: Run CI Failure Bot
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          GEMINI_API_KEY: ${{ secrets.GEMINI_API_KEY }}
+          WORKFLOW_RUN_ID: ${{ github.event.workflow_run.id }}
+          REPOSITORY: ${{ github.repository }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number || '' }}
+        run: python .github/scripts/ci_failure_bot.py
```
[the remark this answers is not in the document]
Comment (Srinath0916) [reply .github/workflows/ci-failure-bot.yml]: @nemesifier , Yes, the workflow will fail loudly now. Added explicit error checking so any script issues will be immediately visible in the GitHub Actions logs.

**Comment by Srinath0916** [comment]
Hi @nemesifier ,

I followed the reusable GitHub utils documentation you linked and aligned both RST files to match that structure. I also validated them locally with docutils, fixed EOF newlines, and confirmed they render correctly.

CI is now failing only on the ReStructuredText formatting step with “file could be reformatted”, and I’m not able to reproduce that behavior locally. It looks like there may be a specific RST formatter or configuration used in CI that I’m missing.

Could you please let me know which tool or command is used to normalize RST files in OpenWISP, or if there’s a preferred way to auto-format these docs before committing? Once I know that, I’ll fix it right away and avoid further commit churn.

Thanks.

File: .github/scripts/ci_failure_bot.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+Analyze the failure and provide your response:
+"""
+
+        try:
+            response = self.model.generate_content(context)
+            return response.text
+        except (ValueError, ConnectionError, Exception) as e:
+            print(f"Error calling Gemini API: {e}")
+            return self.fallback_response()
+
```
Comment (Srinath0916) [inline .github/scripts/ci_failure_bot.py]: 🚫 SKIP

File: .github/workflows/ci-failure-bot.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+      - name: Set up Python
+        uses: actions/setup-python@v6
+        with:
+          python-version: "3.11"
+
+      - name: Install dependencies
+        # Install minimum required versions
+        run: |
+          pip install requests>=2.32.5 PyGithub>=2.0.0 google-genai>=0.2.0
```
Comment (nemesifier) [inline .github/workflows/ci-failure-bot.yml]: here we could add these as a new dependencies in a new extra_requires, eg: `openwisp-utils[github_actions]`, so we would do  something like : `pip install -e .[github_actions]`.

File: docs/index.rst [hunk 7]
```diff
@@ -37,6 +37,7 @@ OpenWISP architecture.
     ./user/metric-collection.rst
     ./user/admin-filters.rst
     ./user/settings.rst
+    ./user/ci-failure-bot.rst
```
Comment (nemesifier) [inline docs/index.rst]: same here

File: docs/developer/reusable-github-utils.rst [hunk 8]
```diff
@@ -56,6 +56,119 @@ times with a 30 second delay between attempts.
...
+  provide specific guidance
+- **Intelligent responses**: Provides direct, actionable feedback based on
+  actual failure context
+- **Comment deduplication**: Updates existing comments instead of creating
+  duplicates
+- **Dependabot exclusion**: Automatically skips dependency update PRs
+- **Fork detection**: Skips external PRs for security
+- **Fallback handling**: Provides basic guidance if AI analysis fails
+
+**Configuration**
```
Comment (nemesifier) [inline docs/developer/reusable-github-utils.rst]: the formatting of this line doesn't make sense as this is bold text and in the next line there's a heading, please use a restructuredtext visualizer to preview this page so you can see the result and edit weird details.

File: openwisp_utils/ci_failure_bot.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+1. **Build Output/Logs:** {build_logs_json}
+2. **YAML Workflow:** {workflow_yaml or "Not available"}
+3. **PR Diff:** {pr_diff_json}
+4. **Project Name:** {project_name}
+5. **Repository:** {repo_url}
+6. **run-qa-checks:** {qa_checks_url}
+7. **runtests:** {runtests_url}
+
+### TASK
+Analyze the provided context to determine why the build failed. Categorize the failure and respond according to the "Tone Guidelines" below.  # noqa: E501
```
Comment (nemesifier) [inline openwisp_utils/ci_failure_bot.py]: I would avoid adding repeated # noqa: E501 which get included in the string, just add line breaks, which also helps readability.

**Comment by Srinath0916** [comment]
Hi @nemesifier ,
I’ve addressed the feedback and updated the PR description to include links to the demo repository and concrete demo PRs showing the bot working on real CI failures across different scenarios.
Please let me know if you’d like me to cover any additional cases.

**Review by nemesifier — CHANGES_REQUESTED** [review]
Let's make sure the bot recommends using our tools:

`openwisp-qa-format` for reformatting code
`./run-qa-checks` for running checks

These are documented in our guidelines. 
If contributors don't use these, they will keep running in loops, reformatting code, getting errors and getting wrong suggestions.

File: setup.py [hunk 10]
```diff
@@ -63,6 +63,11 @@
             "pypandoc~=1.15",
             "pypandoc-binary~=1.15",
         ],
+        "github_actions": [
+            "requests>=2.32.5",
+            "PyGithub>=2.0.0",
+            "google-generativeai>=0.8.0",
+        ],
```
[the remark this answers is not in the document]
Comment (Srinath0916) [reply setup.py]: Thanks for pointing it out
I agree google-generativeai is deprecated and moving to google-genai makes sense. For this PR, I’m keeping it as-is to avoid expanding scope while we validate the bot and workflow.
I’m happy to handle the SDK migration in a follow-up PR once this is merged and tested.

[the remark this answers is not in the document]
Comment (Srinath0916) [reply setup.py]: Thanks! Let’s keep this PR focused for now.
I’m fine handling the google-generativeai -> google-genai migration as a follow-up once the bot is validated in OpenWISP repos, or opening an issue then if it becomes actionable.

**Comment by Srinath0916** [comment]
@nemesifier , done. I’ve updated the bot so that for QA / formatting failures it explicitly recommends the standard OpenWISP workflow (pip install -e .[qa], ./run-qa-checks, openwisp-qa-format).

**Comment by Srinath0916** [comment]
@coderabbitai Coverage dropped due to the new CI failure bot module. Could you point out the key uncovered paths I should add tests for? I'd like to keep the changes minimal.

**Comment by Srinath0916** [comment]
Hi @nemesifier,

I’ve addressed all the review feedback and resolved the CodeRabbit comments.
The bot now consistently recommends only the OpenWISP QA workflow (pip install -e .[qa], openwisp-qa-format, ./run-qa-checks) and handles the relevant edge cases correctly.
I’ve also updated the PR description to include direct links to concrete demo PRs showing the bot working on real CI failures (QA/formatting issues, test failures, and mixed cases), so it’s easier to verify the behavior end-to-end.

Regarding coverage: the drop comes from introducing the new CI failure bot module, which primarily interacts with the GitHub API, workflow_run events, and network paths. These areas are difficult to cover meaningfully without adding artificial or low-value mocks, but the critical logic and edge cases are covered by tests.

The PR should be ready for review whenever you have time. Thanks!

File: runtests.py [hunk 11]
```diff
@@ -16,5 +16,6 @@
     args.insert(1, "test")
     args.insert(2, "test_project")
     args.insert(3, "openwisp_utils.metric_collection")
+    args.insert(4, "openwisp_utils.bots")
```
Comment (nemesifier) [inline runtests.py]: this is wrong, do tests with pytests, look at tests added in #560, look also at the tests for the releaser.

**Review by nemesifier — CHANGES_REQUESTED** [review]
It seems you me that you're going off track @Srinath0916, I see code that doesn't make any sense. You should stop abusing AI tools and go back to the drawing board. I left a couple of comments below.
